### PR TITLE
Feature/add selectors cluster private endpoint

### LINKED ISSUE #609: Add resource selectors for privateEndpointSubnetwork to clusters.container.gcp.upbound.io
### What problem are you facing?
If using `spec.forProvider.subnetworkSelector.matchLabels` to attach to an existing subnet, it would be nice to use label selectors for the `spec.forProvider.privateClusterConfig.privateEndpointSubnetwork`
### What could help solve your problem?
Add resource selectors for `spec.forProvider.privateClusterConfig.privateEndpointSubnetwork`.

### PR DESCRIPTION
**Pull request by bradkwadsworth-mw**
### Description of your changes
Added resource selectors for `spec.forProvider.privateClusterConfig.privateEndpointSubnetwork`

Fixes #609 

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested
Applied to a local Kind cluster
manifest:
```yaml
---
apiVersion: compute.gcp.upbound.io/v1beta1
kind: Network
metadata:
  annotations:
    uptest.upbound.io/timeout: "4800"
    meta.upbound.io/example-id: compute/v1beta1/network
  labels:
    testing.upbound.io/example-name: example_network
  name: example-network
spec:
  forProvider:
    autoCreateSubnetworks: false
---
apiVersion: compute.gcp.upbound.io/v1beta1
kind: Subnetwork
metadata:
  annotations:
    meta.upbound.io/example-id: compute/v1beta1/subnetwork
  labels:
    testing.upbound.io/example-name: example_subnet
  name: gke
spec:
  forProvider:
    ipCidrRange: 10.0.0.0/24
    networkSelector:
      matchLabels:
        testing.upbound.io/example-name: example_network
    region: us-central1
    privateIpGoogleAccess: true
    secondaryIpRange:
      - ipCidrRange: 10.1.0.0/16
        rangeName: pods
      - ipCidrRange: 10.2.0.0/20
        rangeName: services
---
apiVersion: compute.gcp.upbound.io/v1beta1
kind: Router
metadata:
  annotations:
    meta.upbound.io/example-id: compute/v1beta1/router
  labels:
    testing.upbound.io/example-name: example_router
  name: example-router
spec:
  forProvider:
    networkSelector:
      matchLabels:
        testing.upbound.io/example-name: example_network
    region: us-central1
---
apiVersion: compute.gcp.upbound.io/v1beta1
kind: RouterNAT
metadata:
  annotations:
    meta.upbound.io/example-id: compute/v1beta1/routernat
  labels:
    testing.upbound.io/example-name: example_routernat
  name: example-routernat
spec:
  forProvider:
    natIpAllocateOption: AUTO_ONLY
    region: us-central1
    routerSelector:
      matchLabels:
        testing.upbound.io/example-name: example_router
    sourceSubnetworkIpRangesToNat: ALL_SUBNETWORKS_ALL_IP_RANGES
---
apiVersion: container.gcp.upbound.io/v1beta1
kind: Cluster
metadata:
  annotations:
    meta.upbound.io/example-id: container/v1beta1/cluster
  labels:
    testing.upbound.io/example-name: cluster
  name: cluster
spec:
  forProvider:
    addonsConfig:
      - gcePersistentDiskCsiDriverConfig:
          - enabled: true
        gcpFilestoreCsiDriverConfig:
          - enabled: true
        gkeBackupAgentConfig:
          - enabled: true
    binaryAuthorization:
      - evaluationMode: PROJECT_SINGLETON_POLICY_ENFORCE
    datapathProvider: ADVANCED_DATAPATH
    deletionProtection: false
    enableAutopilot: false
    enableIntranodeVisibility: true
    enableL4IlbSubsetting: true
    enableShieldedNodes: true
    gatewayApiConfig:
      - channel: CHANNEL_STANDARD
    initialNodeCount: 1
    ipAllocationPolicy:
      - clusterSecondaryRangeName: pods
        servicesSecondaryRangeName: services
    location: us-central1
    masterAuthorizedNetworksConfig:
      - gcpPublicCidrsAccessEnabled: false
        cidrBlocks:
          - cidrBlock: 10.0.0.0/24
            displayName: bastion
    networkSelector:
      matchLabels:
        testing.upbound.io/example-name: example_network
    networkingMode: VPC_NATIVE
    privateClusterConfig:
      - enablePrivateEndpoint: true
        enablePrivateNodes: true
        masterGlobalAccessConfig:
          - enabled: false
        privateEndpointSubnetworkSelector:
          matchLabels:
            testing.upbound.io/example-name: example_subnet
    releaseChannel:
      - channel: STABLE
    removeDefaultNodePool: true
    subnetworkSelector:
      matchLabels:
        testing.upbound.io/example-name: example_subnet
```
Screenshot 2024-09-03 at 5 25 54 PM

### CODE REVIEW

**Comment by turkenf** [comment]
Hi @bradkwadsworth-mw, thanks for the PR.

For failed CI pipelines, please run `make submodules` and force push the changes.

**Comment by bradkwadsworth-mw** [comment]
I ran `make submodules` and there were no changes.

**Comment by bradkwadsworth-mw** [comment]
@turkenf It looks like the tests are passing now.

**Comment by bradkwadsworth-mw** [comment]
@turkenf Just rebased against current main.

**Comment by turkenf** [comment]
/test-examples="examples/cloudplatform/v1beta1/serviceaccount.yaml"

File: apis/container/v1beta1/zz_cluster_types.go [hunk 1]
```diff
@@ -4603,8 +4603,18 @@ type PrivateClusterConfigObservation struct {
...
 	PrivateEndpointSubnetwork *string `json:"privateEndpointSubnetwork,omitempty" tf:"private_endpoint_subnetwork,omitempty"`
 
+	// Reference to a Subnetwork in compute to populate privateEndpointSubnetwork.
+	// +kubebuilder:validation:Optional
+	PrivateEndpointSubnetworkRef *v1.Reference `json:"privateEndpointSubnetworkRef,omitempty" tf:"-"`
+
+	// Selector for a Subnetwork in compute to populate privateEndpointSubnetwork.
+	// +kubebuilder:validation:Optional
+	PrivateEndpointSubnetworkSelector *v1.Selector `json:"privateEndpointSubnetworkSelector,omitempty" tf:"-"`
+
```
Comment (turkenf) [inline apis/container/v1beta1/zz_cluster_types.go:4617]: I see that these changes are made in PrivateClusterConfigInitParameters in `v1beta2` file, can you please move it from observation to init parameters?

File: apis/container/v1beta1/zz_cluster_types.go [hunk 2]
```diff
@@ -4641,8 +4651,17 @@ type PrivateClusterConfigParameters struct {
 	MasterIPv4CidrBlock *string `json:"masterIpv4CidrBlock,omitempty" tf:"master_ipv4_cidr_block,omitempty"`
 
 	// Subnetwork in cluster's network where master's endpoint will be provisioned.
-	// +kubebuilder:validation:Optional
```
Comment (turkenf) [inline apis/container/v1beta1/zz_cluster_types.go:4644]: Is there a reason you removed this line?

**Review by turkenf — COMMENTED** [review]
Thank you for your effort @bradkwadsworth-mw, I left two comments. And, do you have a specific reason for updating the build submodule, if not would you revert it to the same commit as main?

**Comment by bradkwadsworth-mw** [comment]
@turkenf I think I fixed the issues.

**Comment by turkenf** [comment]
@bradkwadsworth-mw, as I mentioned [here](https://github.com/crossplane-contrib/provider-upjet-gcp/pull/610#pullrequestreview-2357816694): do you have a specific reason for updating the build submodule, if not would you revert it to the same commit as main?

**Comment by turkenf** [comment]
And please squash your commits for a clean commit history.

**Comment by bradkwadsworth-mw** [comment]
@turkenf ok should be good I hope.

**Comment by turkenf** [comment]
Please add the following lines to `PrivateClusterConfigInitParameters` and `PrivateClusterConfigParameters` in the `apis/container/v1beta1/zz_cluster_types.go`  file.
```
	// +crossplane:generate:reference:type=github.com/upbound/provider-gcp/apis/compute/v1beta2.Subnetwork
	// +crossplane:generate:reference:extractor=github.com/upbound/provider-gcp/config/common.SelfLinkExtractor()
```

You just need to copy changes in types go file from `v1beta2` to `v1beta1`

**Comment by bradkwadsworth-mw** [comment]
@turkenf hopefully that should do it.

**Comment by turkenf** [comment]
Could you please run make generate again and push the changes?

File: apis/container/v1beta1/zz_cluster_types.go [hunk 3]
```diff
@@ -4643,6 +4653,14 @@ type PrivateClusterConfigParameters struct {
 	// Subnetwork in cluster's network where master's endpoint will be provisioned.
```
[the remark this answers is not in the document]
Comment (bradkwadsworth-mw) [reply apis/container/v1beta1/zz_cluster_types.go:4653]: I seem to be getting a few `import cycle not allowed` errors when I do a `make generate` is that expected?

[the remark this answers is not in the document]
Comment (turkenf) [reply apis/container/v1beta1/zz_cluster_types.go:4653]: It shouldn't be, I'll check

[the remark this answers is not in the document]
Comment (turkenf) [reply apis/container/v1beta1/zz_cluster_types.go:4653]: Hi @bradkwadsworth-mw, if you see logs like the screenshot below, this is expected, but if the generation fails, it is unexpected.

Screenshot 2024-10-25 at 23 19 08

[the remark this answers is not in the document]
Comment (bradkwadsworth-mw) [reply apis/container/v1beta1/zz_cluster_types.go:4653]: Ok got it, the errors from the screenshot were the same ones I was getting.

**Review by turkenf — APPROVED** [review]
Thank you @bradkwadsworth-mw, LGTM.

**Pull request merged**